Qiskit wrapper

gcattan opened the pull request:
Issue:
https://github.com/pyRiemann/pyRiemann/issues/141

Blocker:
https://github.com/pyRiemann/pyRiemann-qiskit/pull/1

This PR is a "fresh one" but a similar PR with commit history was raised in https://github.com/pyRiemann/pyRiemann-qiskit/pull/3

qbarthelemy commented on .coveragerc:
```diff
@@ -0,0 +1,8 @@
+[report]
```
This file seems already present in #1.
Same remark for several files of this PR.

Anyway, #2 should be rebased after merge of #1.

gcattan replied:
Hello, 
Yes, #1 was set as a blocker for this PR , that is #1 is a base branch for this PR. I will remerge #1 into #2 after #2 is merge to solve conflicts.

qbarthelemy replied:
My advice is to **remove all duplicated files** from PR 2:
- to lighten the diff, avoiding to review many unmodified files;
- to avoid future conflicts when PR 1 will be merged.

In fact, PR 2 needs to add only 4 new files:
```
pyriemann_qiskit/classification.py
tests/test_classification.py
examples/ERP/README.txt
examples/ERP/plot_classify_EEG_quantum.py
```
All other files already exist in PR 1.

gcattan replied:
Hum... there will be likely no merge conflict because this branch is issued from the one of PR 1 and they share the same commit history. Also I need some of the files defined in the underlying architecture, such as __init__.py or setup.py. 
Would it be more comfortable to pursue the review in this draft: https://github.com/gcattan/pyRiemann-qiskit/pull/1
(branch is the same so any modification there will be applicable here)?

qbarthelemy replied:
All these PR are a little bit too messy for me. 
But I prefer continue on this PR, even though the diff is not easy to review.

gcattan replied:
I understand and will take into account your consideration for future PRs. 
Once again, thank you very much for your hard work in reviewing this code. I know this is already a lot of work and I highly appreciate your time and help to improve the quality of this implementation.

qbarthelemy commented on examples/ERP/plot_classify_EEG_quantum.py:
```diff
@@ -0,0 +1,127 @@
+# in the doc pipeline
+__cvxpy__ = True
+try:
+    import cvxpy
```
Why do you need this import, since the example does not use any functions/classes from this package?

gcattan replied:
Oh sorry, I made some comments in the initial PR I forgot to report here. 
Wheel fail to build the cvxpy library in the doc pipeline. Therefore it happens that the cvxpy library is not imported and that Qiskit is not working when building the doc. 
This is a temporal workaround to avoid the pipeline to fail, as explained in the comment above.

qbarthelemy replied:
Ok. IMO, an issue can be opened later to properly solve this problem.

gcattan replied:
Thank you for your answer. I confirm issue #5 was created.

qbarthelemy commented on examples/ERP/plot_classify_EEG_quantum.py:
```diff
@@ -0,0 +1,127 @@
+# (If not, then matrices will be inlined inside the quantum classifier)
+tg = TangentSpace()
+
+# Results will be computed for QuanticSVM versus SKLearnSVM for comparison
```
Yes, can you add the comparison with the classical SVM?

qbarthelemy replied:
Oh. `quantum=False` is the classical SVM, isn't it?

gcattan replied:
Yes :)

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,317 @@
+    This class implements a SKLearn wrapper around Qiskit library.
+    It provides a mean to run classification tasks on a local and
+    simulated quantum computer or a remote and real quantum computer.
+    Difference between simulated and real quantum computer will be that:
+    - There is no noise on a simulated quantum computer (so results are better)
+    - Real quantum computer are quicker than simulator
+    - Real quantum computer tasks are assigned to a queue
+      before being executed on a back-end
```
```suggestion
    Difference between simulated and real quantum computer will be that:

    * There is no noise on a simulated quantum computer (so results are better)
    * Real quantum computer are quicker than simulator
    * Real quantum computer tasks are assigned to a queue
      before being executed on a back-end
```

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,317 @@
+    ----------
+    classes_ : list
+        list of classes.
+    verbose : see above
```
`verbose` is an input parameter, it is not a inner attribute of the class.
Same remark for parameters below.

Moreover, in scikit conventions, attributes have a _ at the end of their names.

gcattan replied:
Please, note that in the constructor: `self.verbose = verbose` (also apply for parameters below).

The naming convention was updated as requested.

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,317 @@
+        self.training_input = {}
+        self.target = target
+        self.quantum = quantum
+        if quantum:
```
Have a look to this excellent comment:
https://github.com/pyRiemann/pyRiemann/pull/122#discussion_r648021748

gcattan replied:
Thank you for sharing the comment, you are right.

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,317 @@
+        else:
+            self.log("Classical SVM will be performed")
+
+    def log(self, *values):
```
The name of private functions must begin by a _ .

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,317 @@
+                                                    seed_transpiler=seed_trans)
+        return self
+
+    def get_params(self, deep=True):
+        # Class is re-instanciated for each fold of a cv pipeline.
+        # Deep copy of the original instance is insure trough this method
+        # and the pending one set_params
+        return {
+            "target": self.target,
+            "qAccountToken": self.qAccountToken,
+            "quantum": self.quantum,
+            "processVector": self.processVector,
+            "verbose": self.verbose,
+            "test_input": self.test_input,
+        }
+
+    def set_params(self, **parameters):
+        for parameter, value in parameters.items():
+            setattr(self, parameter, value)
+        return self
```
I believe that these functions are already inherited. To check.

gcattan replied:
I made some tests with a validation pipeline and had to implement these methods as a workaround. Could it be the naming convention?

gcattan replied:
Yes, I guess it is exactly related to this comment you gave me:
https://github.com/pyRiemann/pyRiemann/pull/122#discussion_r648021748

gcattan replied:
Ok, I got it sorted by passing test_input as a parameter.
I also provided one other test `test_GetSetParams` using a cross-validation pipeline, to insure cloning is working as expected

qbarthelemy commented on tests/test_classification.py:
```diff
@@ -0,0 +1,148 @@
+def test_Quantic_splitTargetAndNonTarget(get_covmats):
+    """Test splitTargetAndNonTarget method of quantum classifiers"""
+    covset = get_covmats(100, 3)
+    labels = np.array([0, 1]).repeat(50)
```
You can use `get_labels` here.

qbarthelemy commented on tests/test_classification.py:
```diff
@@ -0,0 +1,148 @@
+    # if "classical" computation enable,
+    # no provider and backend should be defined
+    q = QuanticSVM(target=1, quantum=False)
+    assert(not q.quantum)
```
You can lighten your tests, removing brackets after `assert`.

qbarthelemy commented on tests/test_classification.py:
```diff
@@ -0,0 +1,148 @@
+
+def test_Quantic_splitTargetAndNonTarget(get_covmats):
+    """Test splitTargetAndNonTarget method of quantum classifiers"""
+    covset = get_covmats(100, 3)
```
Have a look to new tests: numeric parameters are defined in a separate line.
For eg, `n_matrices, n_channels = 100, 3`,
allowing non hardcoded asserts, like: `assert len(xta) == n_matrices//2`

qbarthelemy commented on tests/test_classification.py:
```diff
@@ -0,0 +1,148 @@
+    labels = np.array([0, 1]).repeat(50)
+    q = QuanticSVM(target=1, quantum=False)
+    xta, xnt = q.splitTargetAndNonTarget(covset, labels)
+    assert(len(xta) == 50)
+    # Covariance matrices should be vectorized
+    assert(len(xta[0]) == 3 * 3)
```
This can be testes in one line:
`assert xta.shape == (n_matrices, n_channels * n_channels)`

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,318 @@
+            nbSamples = len(X[0][0])
+        except Exception:
+            nbSamples = 1
+        self.feature_dim = nbSensor * nbSamples
```
To be homogeneous with pyRiemann variable naming, can you use `n_channels` instead of `nbSensor`, and `n_matrices` instead of `nbSamples` ?

See
https://github.com/pyRiemann/pyRiemann/pull/120/files#r646880031
and
https://github.com/pyRiemann/pyRiemann/pull/140#discussion_r714660528

qbarthelemy commented on tests/test_classification.py:
```diff
@@ -0,0 +1,148 @@
+    prediction = q.predict(covset)
+    # Considering the inputs, this probably make no sense to test accuracy.
+    # Instead, we could consider this test as a canary test
+    assert(len(prediction) == len(labels))
```
```suggestion
    assert len(prediction) == len(labels)
```

gcattan commented:
@qbarthelemy Thank you for your time reviewing this PR. 
I have addressed your comment by improving compliance with scikit architecture (removing of set and params methods and better naming of the variables) and in general code readability as suggested.

One test was also added in order to confirm the behavior of the cloning functionality.

I have opened a draft at https://github.com/gcattan/pyRiemann-qiskit/pull/1, so you can better visualize the difference between this branch and the base branch (#1).

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,296 @@
+    --------
+    QuanticBase
+
+    """
```
Can you add 1 or 2 references for QSVM?
```suggestion
    References
    ----------
    .. [1] TODO
    .. [2] TODO
    """
```

qbarthelemy replied:
Same remark for `VQC`.

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,296 @@
+    qAccountToken : string (default:None)
+        If quantum==True and qAccountToken provided,
+        the classification task will be running on a IBM quantum backend
+    processVector : lambda vector: processedVector (default)
```
```suggestion
    processVector : lambda vector (default: lambda v: v)
```

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,296 @@
+        Additional processing on the input vectors. eg: downsampling
+    verbose : bool (default:True)
+        If true will output all intermediate results and logs
+    quantum : Bool (default:True)
```
```suggestion
    quantum : bool (default: True)
```

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,296 @@
+        - If true will run on local or remote backend
+        (depending on qAccountToken value).
+        - If false, will perform classical computing instead
+    test_input : Dictionnary
```
```suggestion
    test_input : dict
```

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,296 @@
+
+    Attributes
+    ----------
+    _classes : list
```
Public attributes have "_" on their right side, protected ones on the left side.

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,296 @@
+        - If false, will perform classical computing instead
+    test_input : Dictionnary
+        Contains vectorized test set for target and non-target classes
+
```
```suggestion

    Notes
    -----
    .. versionadded:: 0.0.0.dev

```

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,296 @@
+    def _additional_setup(self):
+        self._log("There is no additional setup.")
+
+    def fit(self, X, y):
```
Can you add documentation for public functions, describing inputs and outputs?

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,296 @@
+logger.level = logging.INFO
+
+
+class QuanticBase(BaseEstimator, ClassifierMixin):
```
`QuanticBase` seems a little bit too generic name: for eg, what would be the name of a base class for quantic regression?
`QuanticBase` could be renamed `QuanticClassifierBase`, avoiding future conflicts when adding new classes, like `QuanticRegressorBase`.

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,360 @@
+    ----------
+    target : int
+        Label of the target symbol
+    q_account_token : string (default:None)
```
Since `q_account_token` explanation depends on `quantum`, you should describe `quantum` before.

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,360 @@
+    def _split_target_and_non_target(self, X, y):
+        self._log("""[Warning] Spitting target from non target.
+                 Only binary classification is supported.""")
+        n_matrices = len(X[0])
+        try:
+            n_channels = len(X[0][0])
+        except Exception:
+            n_channels = 1
```
You should test the suggestion before acceptation.
```suggestion
        X = np.atleast_2d(X)
        n_matrices, n_channels = X.shape
```

gcattan replied:
Nice! X being a set of matrices, `X = np.atleast_2d(X[0])` should works.

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,360 @@
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_trials, n_channels, n_channels) |
+                     shape (n_trials, n_channels * n_channels)
```
```suggestion
        X : ndarray, shape (n_trials, n_channels, n_channels) | \
                shape (n_trials, n_channels * n_channels)
```

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,360 @@
+                     shape (n_trials, n_channels * n_channels)
+            ndarray of SPD matrices.
+            process_vector argument is required if providing 2d matrices
+        y : ndarray shape (n_trials, 1)
```
Is it really a 2D array?

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,360 @@
+    def _self_calibration(self):
+        X = self._prev_fit_params["X"]
+        y = self._prev_fit_params["y"]
+        test_per = 0.33
```
Looks like a magic number... Interesting to be defined as a parameter?

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,360 @@
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_trials, n_channels, n_channels) |
+                     shape (n_trials, n_channels * n_channels)
```
```suggestion
        X : ndarray, shape (n_trials, n_channels, n_channels) | \
                shape (n_trials, n_channels * n_channels)
```

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,360 @@
+
+        Returns
+        -------
+        pred : array of string, shape (n_trials, 1)
```
Is it really a 2D array?

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,360 @@
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_trials, n_channels, n_channels) |
+                     shape (n_trials, n_channels * n_channels)
```
```suggestion
        X : ndarray, shape (n_trials, n_channels, n_channels) | \
                shape (n_trials, n_channels * n_channels)
```

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,360 @@
+        ----------
+        X : ndarray, shape (n_trials, n_channels, n_channels) |
+                     shape (n_trials, n_channels * n_channels)
+            ndarray of SPD matrices.
```
Doc should be completed: it seems that matrices can be provided in raw format (3D array), or after vectorization (2D).
Is it a basic vectorization? or, is it a vectorization after projection in the tangent space?

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,360 @@
+    q_account_token : see QuanticClassifierBase
+    process_vector : see QuanticClassifierBase
+    verbose : see QuanticClassifierBase
+    parameters : see QuanticClassifierBase
+
+    See Also
```
```suggestion
    parameters : see QuanticClassifierBase

    Notes
    -----
    .. versionadded:: 0.0.1

    See Also
```

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,360 @@
+
+    This class implements SVC on a quantum machine.
+    Note if `quantum` parameter is set to `False`
+    then a classical SVC will be perfomed instead.
+
+    See Also
```
```suggestion
    then a classical SVC will be perfomed instead.

    Notes
    -----
    .. versionadded:: 0.0.1

    See Also
```

qbarthelemy commented on tests/test_classification.py:
```diff
@@ -0,0 +1,176 @@
+    assert np.shape(xnt) == (class_len, n_channels * n_channels)
+    # assert that the method still works
+    # when passing already vectorized matrices
+    covset.reshape(n_matrices, n_channels * n_channels)
+    try:
+        xta, xnt = q._split_target_and_non_target(covset, labels)
+        assert False  # Should never reach this line
+    except Exception:
+        pass
```
I don't really understand this test, since QuanticSVM is supposed to process these vectorized inputs without problem.

```suggestion
    covset.reshape(n_matrices, n_channels * n_channels)
    xta, xnt = q._split_target_and_non_target(covset, labels)
    assert np.shape(xta) == (class_len, n_channels * n_channels)
    assert np.shape(xnt) == (class_len, n_channels * n_channels)
```

All public functions must be tested twice, with 3D inputs, and then with 2D vectorized inputs.

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,381 @@
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_trials, n_channels, n_channels) |
+                     shape (n_trials, n_channels * n_channels)
```
```suggestion
        X : ndarray, shape (n_trials, n_channels, n_channels) | \
                 shape (n_trials, n_channels * n_channels)
```

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,381 @@
+                shape (n_trials, n_channels * n_channels)
+            ndarray of SPD matrices. Matrices can be provided in
+            raw format (3D) or already vectorized (2D) such as after projection
+            into the tangent space. Raw matrices are naively inlined.
```
Due to the symmetry of SPD matrices, naive vectorization gives a redundant output (halft of the vector is a copy of the other part), of length `n_channels * n_channels`. It exposes your classifier to the curse of dimensionality. If you keep only the upper triangular part, your output will be `n_channels * (n_channels + 1) / 2`, and your classifier will be more robust.

But, I think that, for a generic code, vectorization must be made outside of `QuanticClassifierBase`, using a scikit-learn pipeline (and in this pipeline, you can use pyRiemann merthods to vectorize covariance matrices).
Hence, `QuanticClassifierBase` will process only 2D inputs, of shape (n_trials, n_features), making no implicite hypothesis on the nature of inputs.

gcattan replied:
In general, the `process_vector` attribute can cover this use case, by providing something like this:
![image](https://user-images.githubusercontent.com/6229031/135672568-01f6057e-608c-4cbf-8b93-9270be08ab19.png)

But I can see your point. Better to let pyRiemann taking care of this, it will be more error proning.

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,381 @@
+    q_account_token : string (default:None)
+        If quantum==True and q_account_token provided,
+        the classification task will be running on a IBM quantum backend
+    process_vector : lambda function (default: lambda v: v)
```
In the ERP example, `process_vector` is a function. So, this type is more general:
```suggestion
    process_vector : callable (default: lambda v: v)
```

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,381 @@
+    Parameters
+    ----------
+    target : int
+        Label of the target symbol
```
Code has been deeply improve! Well done!

But I think we can go further, because current description and code have been written for the specific BCI use-case "target versus non-target classification".

Can you write the description and the code for any binary classification, without any reference to target and non-target classes?

For example:
- parameter `target` should be renamed `labels`, being a 2-tuple containing the label of each class;
- variables `Xta` and `Xnt` should be renamed into `X_class0` and `X_class1`,
- etc.

This refactoring will probably help to move towards a multi-class classifier.

gcattan replied:
Sure, you are right. 
It makes me think also to open an issue for the multi-class classifier here: 
https://github.com/pyRiemann/pyRiemann-qiskit/issues/6

qbarthelemy replied:
Ok for the multiclass extension in another PR.
But, in this PR, can you at least rename variables, to provide generic binary classifiers, without any reference to ERP classification?

gcattan replied:
Yes, my message was a bit confusing. I confirm I am doing the renaming of the variable here. 
For the multiclass extension, I will need to investigate it a bit, as I never did it with QS. This is why I created another issue.

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,381 @@
+    simulated quantum computer or a remote and real quantum computer.
+    Difference between simulated and real quantum computer will be that:
+
+    * There is no noise on a simulated quantum computer (so results are better)
+    * Real quantum computer are quicker than simulator
+    * Real quantum computer tasks are assigned to a queue
+      before being executed on a back-end
```
```suggestion
    * there is no noise on a simulated quantum computer (so results are better);
    * real quantum computer are quicker than simulator;
    * real quantum computer tasks are assigned to a queue
      before being executed on a back-end.
```

qbarthelemy commented on examples/README.txt:
```diff
@@ -0,0 +1,6 @@
+Examples Gallery
+================
```
Is it possible to add a new example in the gallery, comparing SVM and QSVM on toy datasets?
You can start from this code: https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html

gcattan replied:
Yes, of course. Do you prefer me to implement it in this PR or in a separate issue?

qbarthelemy replied:
Yes, you can add the example in this PR.

gcattan commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,375 @@
+        Indicate the percent of the fitting set that should be kept for
+        testing. The remaining percent are used for training.
+        This is due to the design of quantum classifiers which require
+        training, testing and predicting sets.
```
A predicting set is another testing set, likely belonging to another dataset. Such as in the example you provide me, where the mesh data that are passed to the `predict_proba` methods do not belong to the initial dataset.
![image](https://user-images.githubusercontent.com/6229031/135981115-444952bf-8e20-4755-9efd-db4868ad180c.png)

qbarthelemy replied:
I don't understand what is a prediction set, because in the toy example, `xx` and `yy` are simply processed like another test set. In classical ML, training is independent from testing; and a fitted classifier can be used on multiple different test sets without problem.
Is prediction set a quantum ML specificity?

gcattan commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,375 @@
+        If true will output all intermediate results and logs
+    test_input : dict (default: {}})
+        Contains vectorized test set for the two classes
+    test_per : double (default: 0.33)
```
This setting tackles a particular use case where only training and predicting datasets are available. This typically appears when calling `predict` just after `fit`. 
As a testing set is mandatory for Qiskit (see comment just above), we use available data from `fit` method and split them in two sets: one for training and another for testing.
This particular setting determines the size of the testing set.

qbarthelemy replied:
In sklearn, `predict` is the function to obtain the predictions on a test set. Maybe what you call "prediction set" is simply equivalent to a "test set" (with or without labels, it does not matter).

gcattan commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,375 @@
+from qiskit.circuit.library import ZZFeatureMap, TwoLocal
+from qiskit.aqua import QuantumInstance, aqua_globals
+from qiskit.aqua.quantum_instance import logger
+from qiskit.aqua.algorithms import QSVM, SklearnSVM, VQC
```
I dont think there is a difference as SklearnSVM is based on sklearn binaries:
https://qiskit.org/documentation/_modules/qiskit/aqua/algorithms/classifiers/sklearn_svm/sklearn_svm.html#SklearnSVM

But the interface is the same as QSVM, so we can use the same inputs/outputs format, which makes it convenient to intergrate into the code.

qbarthelemy commented:
@gcattan , do you know this repo?
https://github.com/Qiskit/qiskit-machine-learning

gcattan commented:
Hello, 
Yes, the package qiskit.ml is deprecated in recent versions of qiskit. This package was moved to qiskit-machine-learning.
Do you want me to upgrade qiskit and import qiskit-machine-learning instead of qiskit.ml?

qbarthelemy commented:
No, my point is that qiskit-machine-learning could be a good community to share, review and maintain your quantic classifiers, because, they are a scikit-learn wrapper more than a pyRiemann one.

gcattan commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,370 @@
+        self._log("Feature dimension = ", self._feature_dim)
+        X_class1 = X[y == self.classes_[1]]
+        X_class0 = X[y == self.classes_[0]]
+        processed_class1 = self._apply_process_vector(X_class1)
```
I agree.

gcattan commented:
> No, my point is that qiskit-machine-learning could be a good community to share, review and maintain your quantic classifiers, because, they are a scikit-learn wrapper more than a pyRiemann one.

As I see it, it could be a good starting point to develop some of the idea presented in issue #7 . But this is definitely a very pertinent suggestion and I will ask there too. 
If during this review, you feel that this work is not suitable anymore for pyRiemann repository, please let me know, I will understand :)

gcattan commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,370 @@
+            print("[QClass] ", *values)
+
+    def _apply_process_vector(self, X):
+        return [self.process_vector(x) for x in X]
```
I removed this method as per your above comment.

qbarthelemy commented:
> As I see it, it could be a good starting point to develop some of the idea presented in issue #7 . But this is definitely a very pertinent suggestion and I will ask there too. If during this review, you feel that this work is not suitable anymore for pyRiemann repository, please let me know, I will understand :)

Ok for me to keep your work in this repo!

gcattan commented on tests/test_classification.py:
```diff
@@ -0,0 +1,163 @@
+    """Test _split_classes method of quantum classifiers"""
+    q = QuanticSVM(quantum=False)
+
+    n_matrices, n_channels, n_classes = 100, 3, 2
```
You are right, I agree. Thank you for the suggestion.

qbarthelemy commented on examples/ERP/plot_classify_EEG_quantum.py:
```diff
@@ -0,0 +1,131 @@
+
+
+# ...and dividing the number of remaining elements by two
+class Downsampler(TransformerMixin):
```
Yes, features downsampling generates is rough and generates a lost of classif power.
My advice is to suppress it, and to apply a supervised or unsupervised dimension reduction.

qbarthelemy replied:
`XdawnCovariances(nfilter=1)` is a weird choice, because after this step, your covariance matrices are reduced to scalars. You should use `XdawnCovariances(nfilter=2)` or `=3` for example.

Moreover, `make_pipeline(sf, cov, tg, Downsampler(), qsvm)` is not OK, because `XdawnCovariances` applies the same processing as `ERPCovariances`. It should be `make_pipeline(cov, tg, qsvm)` or `make_pipeline(sf, tg, qsvm)`.

gcattan replied:
On the figure above, I used these pipeline: `sf -> tg -> Downsampler -> qsvm` and `sf -> tg -> PCA -> qsvm`

Although `nfilter=1`, Xdawn should output 2 channels. 
![image](https://user-images.githubusercontent.com/6229031/137458790-fec54c77-3bbc-4563-9f8f-32e538e3d90c.png)
Unfortunately, I cannot go higher in a quantum computer simulator, as `nfilter=2` will result in a tangent vector of size 36, which is not supported by the quantum simulator :
![image](https://user-images.githubusercontent.com/6229031/137459598-8f0881e0-1e77-4696-bbef-eeef198d47a7.png)

Below is a confusion matrix with the same data. Actually, I am surprised, because (on this particular example) it does not perform as well as with the naive dimension reduction:

![no_dim_red_2channels](https://user-images.githubusercontent.com/6229031/137459687-cd6af92b-1016-4087-9295-c752e05e8ba0.png)

I agree with you though that this Downsampling si really rough and at the ends depends on the indices wich are selected (here below an example of Downsampling but with impair values):

![without_pca_indent](https://user-images.githubusercontent.com/6229031/137461254-0729ce42-e135-4b2e-8b6a-f6210a1a3505.png)

gcattan replied:
Ok, I updated the readme as suggested.
Also, I ran a complete [test ](https://github.com/pyRiemann/pyRiemann-qiskit/files/7358211/compare_dim_red.txt) on "vis left vs right" with a 5-fold cross-validation pipeline on all trials. The size of the feature was reduced from 10 to 5. Results are below (top line is SVC and bottom line is QSVM, score is balanced accuracy):
![dim_red](https://user-images.githubusercontent.com/6229031/137600886-58555b5c-0071-4397-8565-e193073ada05.png)

qbarthelemy commented on examples/ERP/plot_classify_EEG_quantum.py:
```diff
@@ -0,0 +1,131 @@
+    names = ['0(quantum)', '1(quantum)'] if quantum else ['0', '1']
+    cm = confusion_matrix(y_pred, y_test)
+    ConfusionMatrixDisplay(cm, display_labels=names).plot()
+    plt.show()
```
Ok. Can you paste in a comment the figures obtained on your computer for both examples?

gcattan replied:
Below is the output of the toys dataset example (fixed random state, no kfolds):

![toys_dataset](https://user-images.githubusercontent.com/6229031/137316247-72f803c7-a34f-46d7-af7d-b7be0f8b2d23.png)

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,354 @@
+        X_class0 = X[y == self.classes_[0]]
+        return (X_class1, X_class0)
+
+    def _additional_setup(self, feature_dim):
```
What is the interest to have `feature_dim` in parameter?

qbarthelemy commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,354 @@
+        the classification task will be running on a IBM quantum backend
+    verbose : bool (default:True)
+        If true will output all intermediate results and logs
+    test_input : dict (default: {}})
```
```suggestion
    test_input : dict (default: {})
```

gcattan commented:
Hello, 
Thank you for your useful comments. I removed the `run` method and instead used directly the `predict`, `train` and `test` methods - and therefore doing the training in the fit method and removed the `test_per` parameter.
I am still working on the example to find appropriate filtering (and not too much time-consuming) for the ERP example, and running once again the toys_dataset (as I changed the predict_proba method).
I will then post the pictures in the slack as you mentioned. 

I got also a response from the `qiskit-ml` mainteners. They said that as compared to qiskit, qiskit-ml was recently refactored to support SKlearn (but admittedly not tested it with sklearn pipeline). I guess a follow-up issue to the current review would be to replace qiskit.aqua by qiskit-ml, doing the appropriate testing, which in sum is a good news, as it will allow to focus on more pyRiemann work :)

qbarthelemy commented on examples/toys_dataset/plot_classifier_comparison.py:
```diff
@@ -0,0 +1,160 @@
+            get_qiskit_dataset()
+           ]
+
+figure = plt.figure(figsize=(20, 9))
```
```suggestion
figure = plt.figure(figsize=(15, 9))
```

qbarthelemy commented on examples/toys_dataset/plot_classifier_comparison.py:
```diff
@@ -0,0 +1,160 @@
+classifiers on toys datasets.
+
+"""
+# Code source: Gaël Varoquaux
+#              Andreas Müller
+# Modified for documentation by Jaques Grobler
```
```suggestion
# https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html
```

qbarthelemy replied:
```suggestion
# Code source: https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html
```

qbarthelemy commented on examples/ERP/plot_classify_EEG_quantum.py:
```diff
@@ -0,0 +1,159 @@
+# the number of elements inside the correlation matrices
+# Thus we reduce elements number by using restrictive spatial filtering
+sf = XdawnCovariances(nfilter=1)
+csp = CSP(nfilter=1)
```
Seems not used.

qbarthelemy commented on examples/ERP/plot_classify_EEG_quantum.py:
```diff
@@ -0,0 +1,159 @@
+# Decoding in tangent space with a quantum classifier
+
+# Time complexity of quantum algorithm depends on the number of trials and
+# the number of elements inside the correlation matrices
```
```suggestion
# the number of elements inside the covariance matrices
```

qbarthelemy commented on examples/ERP/plot_classify_EEG_quantum.py:
```diff
@@ -0,0 +1,159 @@
+class0 = X[y == 0]
+class1 = X[y == 1]
+
+X = np.concatenate((class0[:50],class1[:50]))
+y = np.concatenate(([0] * 50, [1] * 50))
+
+
```
```suggestion
n_trials = 50
X = np.concatenate((class0[:n_trials], class1[:n_trials]))
y = np.concatenate(([0] * n_trials, [1] * n_trials))
```

qbarthelemy commented on examples/toys_dataset/plot_classifier_comparison.py:
```diff
@@ -0,0 +1,159 @@
+
+if __cvxpy__:
+    classifiers.append(QuanticVQC())
+    classifiers.append(QuanticSVM(quantum=False))
```
Ok, but what are the hyper-parameters for ` QuanticSVM(quantum=False)`?
You should use the same as SVC.

And when quantum is True, what are the hyper-parameters of the QuantumKernel?

gcattan replied:
For both there is something like this:
- transform the labels 0/1 into -1/1
- get the kernel matrix:
   - for `quantum=False`, that is call `sklearn.rbf_kernel` with `gamma = None` (hence default is 1/nb_features)
   - for `quantum=True`, that is call `QSVM.get_kernel_matrix` with `feature_map=ZZFeatureMap` and `enforce_spd=True` (set negative eigen values to zero if kernel matrix is not positive semi-definite) as parameters
- optimize the kernel matrix, with similar parameters:
   -  number of iteration=500
   -  normalization of the output = L2 Norm of y
   -  L2 norm regularization factor = 0.001 (C?)
  
I changed `gamma = auto` (i.e 1/n_feature according to the doc), and `C = 0.001`. The figure below show how `QuanticSVM(quantum=False)` compare to `SVC(gamma='auto', C = 0.001)`:

![image](https://user-images.githubusercontent.com/6229031/138146006-1dcc5ccb-aa0a-4d04-93a1-f0eb5fc9383c.png)

gcattan commented on pyriemann_qiskit/classification.py:
```diff
@@ -0,0 +1,369 @@
+            print("[QClass] ", *values)
+
+    def _split_classes(self, X, y):
+        self._log("""[Warning] Splitting first class from second class.
+                 Only binary classification is supported.""")
```
Comment deleted as suggested.

sylvchev commented:
Thanks @gcattan, we could merge this PR. Thanks a lot @qbarthelemy for your careful reading!
I'll take care of putting the docs on RTD.
